File: apps/android/app/src/main/java/com/voiid/app/main/clips/ClipFaceSprites.kt
package com.voiid.app.main.clips

import android.content.Context
import android.graphics.BitmapFactory
import androidx.compose.ui.graphics.ImageBitmap
import androidx.compose.ui.graphics.asImageBitmap

/**
 * Photoreal filter art, and where each piece sits on a face.
 *
 * Replaces the hand-drawn vector paths. Those could describe a shape but never a
 * material - flat fills have no way to say "fur" or "gold" - so every filter read as a
 * cartoon no matter how the placement was tuned.
 *
 * ## The placement model
 *
 * Everything is expressed in units of **D**, the interocular distance, because D is the
 * one measurement that scales with the face and is independent of camera resolution,
 * distance and lens. A layer says how wide it is, how far up the face's own axis it sits,
 * and how far out to the side - never anything in pixels.
 *
 * Height is NOT specified: it comes from the bitmap's aspect ratio. The old 512x512
 * sprite-box convention is gone with the vector art, and with it the need to keep art and
 * placement constants agreeing about where the edges of an invisible box were.
 *
 * Reference proportions, measured from the device:
 *   top of head  ~1.9 D above the eye line   (NOT the ML Kit box, which stops at the brow)
 *   head width   ~2.6 D
 *   nose base    ~0.55 D below the eye line
 *   mouth        ~1.0 D below the eye line
 */

enum class FxAnchor { EYES, NOSE, MOUTH }

/**
 * One drawn piece.
 *
 * @param widthD    width in units of D; height follows the bitmap's aspect
 * @param riseD     distance along the face's UP axis from the anchor (negative = down)
 * @param outD      distance along the face's RIGHT axis; a [mirrored] layer draws at both
 *                  -outD and +outD
 * @param mirrored  true for pairs: one bitmap, drawn twice, flipped on the second copy, so
 *                  the two sides can never drift apart the way two AI-generated ears would
 * @param flipBase  flips the un-mirrored copy, for source art that faces the wrong way
 * @param wobble    carries the ear spring, so only soft things bounce
 * @param openMouth if set, the layer only draws once mouthOpenness passes this
 * @param pivotX    where the anchor sits INSIDE the bitmap, as a fraction of its size
 * @param pivotY    likewise vertically. Measured, not guessed: a tiger mask's eye holes
 *                  are at 0.42 of its height and a tiger muzzle's nose pad at 0.22, so
 *                  pinning either bitmap by its centre hangs the feature off the face.
 */
data class FxLayer(
    val asset: String,
    val anchor: FxAnchor,
    val widthD: Float,
    val riseD: Float = 0f,
    val outD: Float = 0f,
    val mirrored: Boolean = false,
    val flipBase: Boolean = false,
    val wobble: Boolean = false,
    val openMouth: Float? = null,
    val pivotX: Float = 0.5f,
    val pivotY: Float = 0.5f,
)

object ClipFaceArt {

    /**
     * Layers are drawn in list order, so things meant to sit behind the face - ears,
     * horns, a hat - come before anything that sits on it.
     */
    val layers: Map<ClipFaceEffect, List<FxLayer>> = mapOf(
        ClipFaceEffect.NONE to emptyList(),

        ClipFaceEffect.DOG to listOf(
            FxLayer("dog_ear_left", FxAnchor.EYES, widthD = 0.95f, riseD = 1.35f,
                    outD = 0.95f, mirrored = true, wobble = true),
            FxLayer("dog_nose", FxAnchor.NOSE, widthD = 0.82f, riseD = -0.04f,
                    pivotY = 0.47f),
            // Only when the mouth actually opens, which is the whole joke.
            FxLayer("dog_tongue", FxAnchor.MOUTH, widthD = 0.50f, riseD = -0.42f,
                    openMouth = 0.15f, pivotY = 0.10f),
        ),

        ClipFaceEffect.TIGER to listOf(
            FxLayer("tiger_ear_left", FxAnchor.EYES, widthD = 0.85f, riseD = 1.28f,
                    outD = 0.86f, mirrored = true, wobble = true),
            // The mask and the muzzle together make the face, so both are pinned by their
            // own features - eye holes at 0.42, nose pad at 0.22 - not by their centres.
            FxLayer("tiger_stripes", FxAnchor.EYES, widthD = 2.50f, riseD = 0.0f,
                    pivotY = 0.42f),
            FxLayer("tiger_nose", FxAnchor.NOSE, widthD = 1.60f, riseD = -0.02f,
                    pivotY = 0.22f),
        ),

        ClipFaceEffect.PARTY to listOf(
            FxLayer("party_hat", FxAnchor.EYES, widthD = 1.15f, riseD = 2.55f),
        ),

        ClipFaceEffect.CYBER to listOf(
            FxLayer("cyber_visor", FxAnchor.EYES, widthD = 2.35f, riseD = 0.06f),
        ),

        ClipFaceEffect.BUNNY to listOf(
            FxLayer("bunny_ear_left", FxAnchor.EYES, widthD = 0.46f, riseD = 2.10f,
                    outD = 0.48f, mirrored = true, wobble = true),
            FxLayer("bunny_nose", FxAnchor.NOSE, widthD = 1.10f, riseD = -0.02f,
                    pivotY = 0.29f),
        ),

        ClipFaceEffect.KOALA to listOf(
            FxLayer("koala_ear_left", FxAnchor.EYES, widthD = 1.15f, riseD = 1.45f,
                    outD = 1.20f, mirrored = true, wobble = true),
            FxLayer("koala_nose", FxAnchor.NOSE, widthD = 0.86f, riseD = -0.02f,
                    pivotY = 0.58f),
        ),

        ClipFaceEffect.CAT to listOf(
            FxLayer("cat_ear_left", FxAnchor.EYES, widthD = 0.88f, riseD = 1.50f,
                    outD = 0.88f, mirrored = true, flipBase = true, wobble = true),
            // A muzzle, not a nose: it has to reach the mouth, so it is wider than the
            // nose pad it is pinned by.
            FxLayer("cat_nose", FxAnchor.NOSE, widthD = 1.50f, riseD = -0.02f,
                    pivotY = 0.44f),
            FxLayer("cat_whiskers", FxAnchor.NOSE, widthD = 2.45f, riseD = -0.02f),
        ),

        ClipFaceEffect.SUNGLASSES to listOf(
            FxLayer("sunglasses", FxAnchor.EYES, widthD = 2.35f, riseD = 0.06f),
        ),

        // Sized to the head (~2.6 D wide) rather than to the face box, and dropped so the
        // band beds into the hair instead of hovering above it.
        ClipFaceEffect.CROWN to listOf(
            FxLayer("crown", FxAnchor.EYES, widthD = 2.85f, riseD = 1.95f),
        ),

        // The source ring is 347x116 (aspect 2.99). It was extracted clipped twice - the
        // crop box cut through the right of the ring - which made it look both narrower
        // and taller than it is, so the old width was tuned against a broken sprite.
        ClipFaceEffect.HALO to listOf(
            FxLayer("halo", FxAnchor.EYES, widthD = 2.45f, riseD = 2.50f),
        ),

        // flipBase: the source horn curves one way, and drawn unflipped on the left the
        // pair curled inward at each other like ram's horns.
        ClipFaceEffect.DEVIL to listOf(
            FxLayer("devil_horn_left", FxAnchor.EYES, widthD = 0.60f, riseD = 1.85f,
                    outD = 0.60f, mirrored = true, flipBase = true),
        ),
    )
}

/**
 * Loads and holds the filter bitmaps.
 *
 * Decoding happens once, off the draw path. A miss returns null and the layer is skipped
 * rather than throwing - a filter with one missing piece is still usable, and a camera
 * that crashes is not.
 */
object ClipFaceAssets {

    private val cache = HashMap<String, ImageBitmap?>()
    private val lock = Any()

    @Volatile
    var ready: Boolean = false
        private set

    fun get(name: String): ImageBitmap? = synchronized(lock) { cache[name] }

    /** Call off the main thread. Safe to call more than once. */
    fun preload(context: Context) {
        val names = ClipFaceArt.layers.values.flatten().map { it.asset }.distinct()
        val loaded = HashMap<String, ImageBitmap?>(names.size)
        for (n in names) {
            loaded[n] = runCatching {
                context.assets.open("filters/$n.png").use { s ->
                    BitmapFactory.decodeStream(s)?.asImageBitmap()
                }
            }.getOrNull()
        }
        synchronized(lock) {
            cache.putAll(loaded)
            ready = true
        }
    }
}

File: apps/ios/Voiid/Voiid/Main/Clips/ClipFaceSprites.swift
//
//  ClipFaceSprites.swift
//  Voiid
//
//  Photoreal filter art and its placement. The Swift half of the same table Android
//  carries in ClipFaceSprites.kt, against the same 19 PNGs — the two platforms must
//  agree on where a dog's ear goes, so the numbers live in one shape on both.
//
//  Everything is expressed in units of D, the interocular distance, which is the only
//  measurement that scales with the face and is independent of camera resolution, lens
//  and distance. Height is never specified: it comes from each bitmap's aspect ratio.
//
//  Reference proportions, measured on device:
//    top of head  ~1.9 D above the eye line   (NOT the detector's box, which stops at the brow)
//    head width   ~2.6 D
//    nose base    ~0.55 D below the eye line
//    mouth        ~1.0 D below the eye line
//

import CoreImage
import UIKit

enum FxAnchor {
    case eyes
    case nose
    case mouth
}

/// One drawn piece.
///
/// - `pivotX` / `pivotYFromTop`: where the anchor sits INSIDE the bitmap, as a fraction.
///   Measured, not guessed — the tiger mask's eye holes are at 0.42 of its height and the
///   tiger muzzle's nose pad at 0.22, so pinning either by its centre hangs the feature
///   off the face. Stated from the TOP to match the Android table; Core Image's y axis
///   points up, so the renderer converts.
/// - `mirrored`: one bitmap, drawn twice, flipped on the far side. Two ears cut from one
///   sprite can never drift apart.
struct FxLayer {
    let asset: String
    let anchor: FxAnchor
    let widthD: CGFloat
    var riseD: CGFloat = 0
    var outD: CGFloat = 0
    var mirrored: Bool = false
    var flipBase: Bool = false
    var wobble: Bool = false
    var openMouth: CGFloat? = nil
    var pivotX: CGFloat = 0.5
    var pivotYFromTop: CGFloat = 0.5
}

enum ClipFaceArt {

    /// Drawn in order, so anything meant to sit behind the face comes first.
    static let layers: [ClipFaceEffect: [FxLayer]] = [
        .none: [],

        .dog: [
            FxLayer(asset: "dog_ear_left", anchor: .eyes, widthD: 0.95, riseD: 1.35,
                    outD: 0.95, mirrored: true, wobble: true),
            FxLayer(asset: "dog_nose", anchor: .nose, widthD: 0.82, riseD: -0.04,
                    pivotYFromTop: 0.47),
            // Only once the mouth actually opens, which is the whole joke.
            FxLayer(asset: "dog_tongue", anchor: .mouth, widthD: 0.50, riseD: -0.42,
                    openMouth: 0.15, pivotYFromTop: 0.10),
        ],

        .tiger: [
            FxLayer(asset: "tiger_ear_left", anchor: .eyes, widthD: 0.85, riseD: 1.28,
                    outD: 0.86, mirrored: true, wobble: true),
            // Mask and muzzle together make the face, each pinned by its own feature.
            FxLayer(asset: "tiger_stripes", anchor: .eyes, widthD: 2.50, riseD: 0,
                    pivotYFromTop: 0.42),
            FxLayer(asset: "tiger_nose", anchor: .nose, widthD: 1.60, riseD: -0.02,
                    pivotYFromTop: 0.22),
        ],

        .party: [
            FxLayer(asset: "party_hat", anchor: .eyes, widthD: 1.15, riseD: 2.55),
        ],

        .cyber: [
            FxLayer(asset: "cyber_visor", anchor: .eyes, widthD: 2.35, riseD: 0.06),
        ],

        .bunny: [
            FxLayer(asset: "bunny_ear_left", anchor: .eyes, widthD: 0.46, riseD: 2.10,
                    outD: 0.48, mirrored: true, wobble: true),
            FxLayer(asset: "bunny_nose", anchor: .nose, widthD: 1.10, riseD: -0.02,
                    pivotYFromTop: 0.29),
        ],

        .koala: [
            FxLayer(asset: "koala_ear_left", anchor: .eyes, widthD: 1.15, riseD: 1.45,
                    outD: 1.20, mirrored: true, wobble: true),
            FxLayer(asset: "koala_nose", anchor: .nose, widthD: 0.86, riseD: -0.02,
                    pivotYFromTop: 0.58),
        ],

        .cat: [
            FxLayer(asset: "cat_ear_left", anchor: .eyes, widthD: 0.88, riseD: 1.50,
                    outD: 0.88, mirrored: true, flipBase: true, wobble: true),
            // A muzzle, not a nose: it has to reach the mouth.
            FxLayer(asset: "cat_nose", anchor: .nose, widthD: 1.50, riseD: -0.02,
                    pivotYFromTop: 0.44),
            FxLayer(asset: "cat_whiskers", anchor: .nose, widthD: 2.45, riseD: -0.02),
        ],

        .sunglasses: [
            FxLayer(asset: "sunglasses", anchor: .eyes, widthD: 2.35, riseD: 0.06),
        ],

        // Sized to the head (~2.6 D) rather than the face box, and low enough that the
        // band beds into the hair instead of hovering above it.
        .crown: [
            FxLayer(asset: "crown", anchor: .eyes, widthD: 2.85, riseD: 1.95),
        ],

        .halo: [
            FxLayer(asset: "halo", anchor: .eyes, widthD: 2.45, riseD: 2.50),
        ],

        // flipBase: the source horn curves one way, and drawn unflipped on the left the
        // pair curled inward at each other like ram's horns.
        .devil: [
            FxLayer(asset: "devil_horn_left", anchor: .eyes, widthD: 0.60, riseD: 1.85,
                    outD: 0.60, mirrored: true, flipBase: true),
        ],
    ]
}

/// Loads and holds the filter bitmaps.
///
/// A miss returns nil and the layer is skipped rather than trapping: a filter with one
/// missing piece is still usable, a camera that crashes is not.
enum ClipFaceAssets {

    private static var cache: [String: CIImage] = [:]
    private static let lock = NSLock()

    static func image(named name: String) -> CIImage? {
        lock.lock()
        if let hit = cache[name] { lock.unlock(); return hit }
        lock.unlock()

        // The PNGs live in Resources/Filters, which is inside a synchronized folder, so
        // they land at the bundle root.
        guard let url = Bundle.main.url(forResource: name, withExtension: "png",
                                        subdirectory: "Filters")
                ?? Bundle.main.url(forResource: name, withExtension: "png"),
              let data = try? Data(contentsOf: url),
              let ui = UIImage(data: data),
              let cg = ui.cgImage
        else { return nil }

        let ci = CIImage(cgImage: cg)
        lock.lock(); cache[name] = ci; lock.unlock()
        return ci
    }

    /// Warms the cache off the render path.
    static func preload() {
        let names = Set(ClipFaceArt.layers.values.flatMap { $0 }.map { $0.asset })
        for n in names { _ = image(named: n) }
    }
}
